plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergBucketFunction.java

```diff
@@ -0,0 +1,140 @@
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.plugin.iceberg.PartitionTransforms.getColumnTransform;
+import static io.trino.spi.function.InvocationConvention.InvocationArgumentConvention.NEVER_NULL;
+import static io.trino.spi.function.InvocationConvention.InvocationReturnConvention.FAIL_ON_NULL;
+import static io.trino.spi.function.InvocationConvention.simpleConvention;
+import static io.trino.spi.type.TypeUtils.NULL_HASH_CODE;
+import static java.util.Objects.requireNonNull;
+
+public class IcebergBucketFunction
+        implements BucketFunction
+{
+    private final int bucketCount;
+
+    private final List<PartitionColumn> partitionColumns;
+    private final List<MethodHandle> hashCodeInvokers;
+
+    public IcebergBucketFunction(
+            TypeOperators typeOperators,
+            PartitionSpec partitionSpec,
+            List<IcebergColumnHandle> partitioningColumns,
+            int bucketCount)
+    {
+        requireNonNull(partitionSpec, "partitionSpec is null");
+        checkArgument(!partitionSpec.isUnpartitioned(), "empty partitionSpec");
+        checkArgument(bucketCount > 0, "Invalid bucketCount: %s", bucketCount);
+        this.bucketCount = bucketCount;
+
+        Map<Integer, Integer> fieldIdToInputChannel = new HashMap<>();
+        for (int i = 0; i < partitioningColumns.size(); i++) {
+            fieldIdToInputChannel.put(partitioningColumns.get(i).getId(), i);
+        }
+        partitionColumns = partitionSpec.fields().stream()
+                .map(field -> {
+                    Integer channel = fieldIdToInputChannel.get(field.sourceId());
+                    checkArgument(channel != null, "partition field not found: %s", field);
+                    Type inputType = partitioningColumns.get(channel).getType();
+                    ColumnTransform transform = getColumnTransform(field, inputType);
+                    return new PartitionColumn(channel, transform.getValueTransform(), transform.getType());
+                })
+                .collect(toImmutableList());
+        hashCodeInvokers = partitionColumns.stream()
+                .map(PartitionColumn::getResultType)
+                .map(type -> typeOperators.getHashCodeOperator(type, simpleConvention(FAIL_ON_NULL, NEVER_NULL)))
+                .collect(toImmutableList());
+    }
+
+    @Override
+    public int getBucket(Page page, int position)
+    {
+        long hash = 0;
+
+        for (int i = 0; i < partitionColumns.size(); i++) {
+            PartitionColumn partitionColumn = partitionColumns.get(i);
+            Block block = page.getBlock(partitionColumn.getSourceChannel());
+            Object value = partitionColumn.getValueTransform().apply(block, position);
+            long valueHash = hashValue(hashCodeInvokers.get(i), value);
+            hash = (31 * hash) + valueHash;
+        }
+
+        return (int) ((hash & Long.MAX_VALUE) % bucketCount);
```

Comment:
i followed @raunaqmorarka 's code from `HivePartitionedBucketFunction` here.

i can replace `hash % bucketCount` with
`(int) ((Integer.toUnsignedLong(Long.hashCode(hash)) * partitionCount) >>> 32)`

but i would not know why i am doing this. since there are other code places that are supposedly going to benefit from that, @raunaqmorarka please file an issue about this and cc @lukasz-stec on it.

i am ignoring this for now